groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+
+    if (d_memoryMap_p == 0) {
+        return 0;
+    }
+
+    bsl::size_t   count = 0;
+    bsl::uint32_t head  = *d_head_p;
+
+    while (true) {
+        NTCO_IORING_READER_BARRIER();
+
+        bsl::uint32_t tail = *d_tail_p;
+
+        if (head == tail) {
+            break;
+        }
+
+        bsl::uint32_t mask  = *d_mask_p;
+        bsl::uint32_t index = head & mask;
+
+        NTCO_IORING_LOG_COMPLETION_QUEUE_POPPING(index);
+
+        result[count] =
+            *reinterpret_cast<ntco::IoRingCompletion*>(&d_entryArray[index]);
+        ++count;
+
+        ++head;
+
+        if (count == capacity) {
+            break;
+        }
+    }
+
+    *d_head_p = head;
+
+    NTCO_IORING_WRITER_BARRIER();
+
+    return count;
+}
+
+void IoRingCompletionQueue::unmap()
+{
+    LockGuard guard(&d_mutex);
+
+    if (d_memoryMap_p != 0) {
+        int rc = ::munmap(d_memoryMap_p,
+                          d_params.completionQueueOffsetToCQEs() +
+                              d_params.completionQueueCapacity() *
+                                  sizeof(ntco::IoRingCompletion));
+        BSLS_ASSERT(rc == 0);
+
+        d_memoryMap_p = 0;
+    }
+}
+
+bsl::uint32_t IoRingCompletionQueue::head() const
+{
+    LockGuard guard(&d_mutex);
+
+    if (d_memoryMap_p != 0) {
+        NTCO_IORING_READER_BARRIER();
+        return *d_head_p;
+    }
+    else {
+        return 0;
+    }
+}
+
+bsl::uint32_t IoRingCompletionQueue::tail() const
+{
+    LockGuard guard(&d_mutex);
+
+    if (d_memoryMap_p != 0) {
+        NTCO_IORING_READER_BARRIER();
+        return *d_tail_p;
+    }
+    else {
+        return 0;
+    }
+}
+
+bsl::uint32_t IoRingCompletionQueue::capacity() const
+{
+    return d_params.completionQueueCapacity();
+}
+
+IoRingDevice::IoRingDevice(bsl::size_t       queueDepth,
+                           bslma::Allocator* basicAllocator)
+: d_ring(-1)
+, d_submissionQueue()
+, d_completionQueue()
+, d_probe()
+, d_params()
+, d_allocator_p(bslma::Default::allocator(basicAllocator))
+{
+    NTCI_LOG_CONTEXT();
+
+    ntsa::Error error;
+    int         rc;
+
+#if NTCO_IORING_RLIMIT_MEMLOCK
+    {
+        rlimit new_limit = {0};
+        rlimit old_limit = {0};
+
+        new_limit.rlim_cur = 1024 * 64;
+        new_limit.rlim_max = 1024 * 64;
+
+        int rc = prlimit(getpid(), RLIMIT_MEMLOCK, &new_limit, &old_limit);
+        if (rc != 0) {
+            ntsa::Error error(errno);
+            NTCO_IORING_LOG_RLIMIT_MEMLOCK_FAILURE(error);
+        }
+
+        NTCO_IORING_LOG_RLIMIT_MEMLOCK_RESULT(old_limit);
+    }
+#endif
+
+    BSLS_ASSERT_OPT(queueDepth <= bsl::numeric_limits<bsl::uint32_t>::max());
+
+    d_ring = ntco::IoRingUtil::setup(queueDepth, &d_params);
+    if (d_ring < 0) {
+        ntsa::Error error(errno);
+        NTCO_IORING_LOG_SETUP_FAILURE(error);
+        NTCCFG_ABORT();
+    }
+    BSLS_ASSERT_OPT(d_ring > 0);
+
+    NTCO_IORING_LOG_CREATED(d_ring);
+
+    rc = ntco::IoRingUtil::probe(d_ring, &d_probe);
```

Comment:
Maybe it can be a good idea to validate that operations like `SENDMSG` (ano other 100% required) are supported.

Reply:
This is a good idea but will require a non-trivial amount of work. I've created https://github.com/bloomberg/ntf-core/issues/67 for this.